### PR TITLE
ADEN-8997 Scroll speed tracking

### PR DESCRIPTION
**Pull request by bckatarzyna**

### CODE REVIEW

File: src/ad-tracking/scroll-tracker.ts [hunk 1]
```diff
@@ -0,0 +1,63 @@
+import { context, events, eventService, scrollSpeedCalculator } from '@ad-engine/core';
+
+class ScrollTracker {
+	scrollSpeedTrackingStarted: boolean;
+	timer: NodeJS.Timer;
+
+	constructor() {
```
Comment (fraszczakszymon) [inline src/ad-tracking/scroll-tracker.ts]: constructor seems not needed. Can we set those values 3 lines above?

File: src/ad-tracking/scroll-tracker.ts [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+	 */
+	initScrollSpeedTracking(applicationAreaClass: string): void {
+		if (!context.get('options.scrollSpeedTracking')) {
+			return;
+		}
+
+		const applicationArea = document.getElementsByClassName(applicationAreaClass)[0];
+
+		this.scrollSpeedTrackingStarted = false;
+		applicationArea.addEventListener('touchstart', () => this.trackScrollSpeedToDW());
```
Comment (fraszczakszymon) [inline src/ad-tracking/scroll-tracker.ts]: It may throw an error (`applicationArea` can be undefined)

Comment (fraszczakszymon) [reply src/ad-tracking/scroll-tracker.ts]: What about SPA? Would below code solve the problem?
```suggestion
		applicationArea.addEventListener('touchstart', () => this.trackScrollSpeedToDW(), { once: true });
```

File: src/ad-tracking/scroll-tracker.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+	}
+
+	/**
+	 * Track scrollY to DW in in three 2s-periods
+	 */
+	trackScrollSpeedToDW(): void {
```
Comment (fraszczakszymon) [inline src/ad-tracking/scroll-tracker.ts]: It doesnt really track to DW. What about `dispatchScrollSpeedEvents`?

File: src/ad-tracking/scroll-tracker.ts (hunk 2 above)
Comment (fraszczakszymon) [reply src/ad-tracking/scroll-tracker.ts]: I just found reset method :P. Anyway "once" would be nice here too

File: src/ad-tracking/scroll-tracker.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+			return;
+		}
+
+		const applicationArea = document.getElementsByClassName(applicationAreaClass)[0];
+
+		this.scrollSpeedTrackingStarted = false;
+		applicationArea.addEventListener('touchstart', () => this.trackScrollSpeedToDW());
+	}
+
+	/**
```
Comment (fraszczakszymon) [inline src/ad-tracking/scroll-tracker.ts]: Aren't we removing jsdocs? :P (in favour of TS)

File: src/ad-tracking/scroll-tracker.ts [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+		}
+
+		const timesToTrack = [0, 2, 4];
+		let startScrollY = 0;
+		this.scrollSpeedTrackingStarted = true;
+
+		timesToTrack.forEach((time) => {
+			this.timer = setTimeout(() => {
+				const scrollY = window.scrollY || window.pageYOffset;
+				eventService.emit(events.TRACK_SCROLL_Y, time, scrollY);
```
Comment (fraszczakszymon) [inline src/ad-tracking/scroll-tracker.ts]: TRACK_SCROLL_Y -> SCROLL_POSITION_CHANGED?

File: src/ad-tracking/scroll-tracker.ts [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+		});
+	}
+
+	/**
+	 * Remove scroll tracking from the page
+	 */
+	resetScrollSpeedTracking(applicationAreaClass: string): void {
+		const applicationArea = document.getElementsByClassName(applicationAreaClass)[0];
+		clearTimeout(this.timer);
+		applicationArea.removeEventListener('touchstart', () => this.trackScrollSpeedToDW());
```
Comment (fraszczakszymon) [inline src/ad-tracking/scroll-tracker.ts]: Does it work? It gets a newly created function in the second argument. Should it be the one added in `addEventListener`?

Comment (fraszczakszymon) [reply src/ad-tracking/scroll-tracker.ts]: What about assigning applicationArea to variable on init, so we don't need to pass class name again to reset method?

File: src/ad-tracking/scroll-tracker.ts (hunk 4 above)
Comment (bckatarzyna) [reply src/ad-tracking/scroll-tracker.ts]: I've copied them from mobile-wiki where there were previously written ;)

**Pull request merged**